### PR TITLE
[pfazure] support sovereign cloud

### PR DESCRIPTION
**Pull request by Stephen1993**
# Description
It can already submit normally:
image

However, the submitted flow has failed to run, and it may be necessary to update the promptflow package used:
image

Please add an informative description that covers that changes made by the pull request and link all relevant issues.

# All Promptflow Contribution checklist:
- [x] **The pull request does not introduce [breaking changes].**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [ ] **I have read the [contribution guidelines](../CONTRIBUTING.md).**
- [ ] **Create an issue and link to the pull request to get dedicated review from promptflow team. Learn more: [suggested workflow](../CONTRIBUTING.md#suggested-workflow).**

## General Guidelines and Best Practices
- [x] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### Testing Guidelines
- [x] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: src/promptflow/promptflow/azure/_restclient/flow_service_caller.py [hunk 1]
```diff
@@ -118,41 +120,20 @@ def __init__(self, workspace, credential, operation_scope, base_url=None, region
...
-            )
-
+        aml_token = get_arm_token(credential=self._credential)
         headers["aml-user-token"] = aml_token
 
     def _get_user_identity_info(self):
         import jwt
 
-        token = self._credential.get_token("https://management.azure.com/.default")
+        token = get_arm_token(credential=self._credential)
```
[the remark this answers is not in the document]
Comment (Stephen1993) [reply src/promptflow/promptflow/azure/_restclient/flow_service_caller.py]: ok, I have already made the necessary modifications。

File: src/promptflow/promptflow/azure/_utils/gerneral.py [hunk 2]
```diff
@@ -3,14 +3,69 @@
...
 
 import jwt
+from promptflow.exceptions import ValidationException
 
 
 def is_arm_id(obj) -> bool:
     return isinstance(obj, str) and obj.startswith("azureml://")
 
 
+def get_aml_token(credential):
```
Comment (wangchao1230) [inline src/promptflow/promptflow/azure/_utils/gerneral.py:13]: looks like we could call get_base_token in this function to reduce dulplicate code?

File: src/promptflow/promptflow/_sdk/operations/_local_azure_connection_operations.py [hunk 3]
```diff
@@ -9,6 +9,7 @@
 from promptflow._sdk._utils import interactive_credential_disabled, is_from_cli, is_github_codespaces, print_red_error
 from promptflow._sdk.entities._connection import _Connection
 from promptflow._utils.logger_utils import get_cli_sdk_logger
+from promptflow.azure._utils.gerneral import get_base_token
```
Comment (wangchao1230) [inline src/promptflow/promptflow/_sdk/operations/_local_azure_connection_operations.py:12]: add a changelog for this change

File: src/promptflow/promptflow/azure/_utils/gerneral.py [hunk 4]
```diff
@@ -3,14 +3,69 @@
...
+        Token with {0} scope can be fetched using credentials.get_token({0})
+        """
+        raise ValidationException(
+            message=msg.format(*azure_ml_scopes),
+        )
+
+    return token
+
+
+def get_base_token(credential):
```
Comment (wangchao1230) [inline src/promptflow/promptflow/azure/_utils/gerneral.py:38]: let's call this get_arm_token

management.azure.com is the rest endpoint for Azure Resource Manager

Comment (Stephen1993) [reply src/promptflow/promptflow/azure/_utils/gerneral.py]: done

File: src/promptflow/promptflow/_sdk/operations/_local_azure_connection_operations.py (hunk 3 above)
Comment (Stephen1993) [reply src/promptflow/promptflow/_sdk/operations/_local_azure_connection_operations.py]: Done, but does the add category feature not need to be added to the changelog?

File: src/promptflow/promptflow/azure/_utils/gerneral.py (hunk 2 above)
Comment (Stephen1993) [reply src/promptflow/promptflow/azure/_utils/gerneral.py]: done

File: src/promptflow/promptflow/_sdk/operations/_local_azure_connection_operations.py [hunk 5]
```diff
@@ -44,12 +44,19 @@ def _client(self):
...
         from azure.identity import DefaultAzureCredential, DeviceCodeCredential
+        from azure.ai.ml._azure_environments import _get_default_cloud_name, EndpointURLS, _get_cloud, AzureEnvironments
 
         if is_from_cli():
             try:
                 # Try getting token for cli without interactive login
-                credential = DefaultAzureCredential()
-                get_base_token(credential=credential)
+                cloud_name = _get_default_cloud_name()
+                if cloud_name != AzureEnvironments.ENV_US_GOVERNMENT:
```
Comment (wangchao1230) [inline src/promptflow/promptflow/_sdk/operations/_local_azure_connection_operations.py:53]: why need special compare here?

Comment (Stephen1993) [reply src/promptflow/promptflow/_sdk/operations/_local_azure_connection_operations.py]: For sovereign, it is necessary to: `authority=authority` and `exclude_shared_token_cache_credential=True`.

Comment (Stephen1993) [reply src/promptflow/promptflow/_sdk/operations/_local_azure_connection_operations.py]: When submitting the flow for local testing, it is written in the following way:
image

image

**Review by wangchao1230 — DISMISSED** [review]

**Review by wangchao1230 — APPROVED** [review]

File: src/promptflow/promptflow/azure/_utils/gerneral.py [hunk 6]
```diff
@@ -3,14 +3,56 @@
...
+    azure_ml_scopes = _resource_to_scopes(resource)
+    token = credential.get_token(*azure_ml_scopes).token
+    # validate token has aml audience
+    decoded_token = jwt.decode(
+        token,
+        options={"verify_signature": False, "verify_aud": False},
+    )
+    if decoded_token.get("aud") != resource:
+        msg = """AAD token with aml scope could not be fetched using the credentials being used.
+            Please validate if token with {0} scope can be fetched using credentials provided to MLClient.
```
Comment (D-W-) [inline src/promptflow/promptflow/azure/_utils/gerneral.py:25]: PFClient?

Comment (D-W-) [reply src/promptflow/promptflow/azure/_utils/gerneral.py:25]: When will this happen?

Comment (Stephen1993) [reply src/promptflow/promptflow/azure/_utils/gerneral.py:25]: This is a reference to the V2 SDK, and I have not encountered this error so far.

Comment (Stephen1993) [reply src/promptflow/promptflow/azure/_utils/gerneral.py:25]: I will change MLClient to PFClient in the next PR to avoid the need for approval again.

**Pull request merged**